### PR TITLE
OneCycleLR with proper budget matching at 100 epochs

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Every prior LR schedule experiment failed because T_max was mismatched with actual epochs or the schedule was too aggressive. OneCycleLR (Smith & Topin, 2019) with `max_lr=0.006`, `epochs=100`, `steps_per_epoch=203` gives proper warmup-then-decay matched to the training budget.

Key difference from CosineAnnealing: OneCycleLR starts LOW (0.00024), ramps to 0.006 over 5% of training, then decays deeply. The warmup stabilizes early training and deep final decay allows finer convergence. All prior OneCycleLR tests (PR #20) used max_lr=0.015-0.02 (way too high) and only 10 epochs — a completely different regime.

## Instructions

In `train.py`, make the following changes:

1. Replace the scheduler creation with:
   ```python
   scheduler = torch.optim.lr_scheduler.OneCycleLR(
       optimizer,
       max_lr=cfg.lr,
       epochs=MAX_EPOCHS,
       steps_per_epoch=len(train_loader),
       pct_start=0.05,        # 5% warmup (~5 epochs)
       anneal_strategy='cos',
       div_factor=25,          # initial_lr = max_lr/25
       final_div_factor=1e4,   # final_lr = initial_lr/1e4
   )
   ```
2. Move `scheduler.step()` inside the training batch loop (OneCycleLR steps per batch, not per epoch):
   ```python
   # After optimizer.step() inside the batch loop:
   scheduler.step()
   ```
3. Remove any `scheduler.step()` call at the epoch level
4. Fix LR logging to use `scheduler.get_last_lr()[0]`
5. Keep all other params: lr=0.006, MAX_EPOCHS=100, bs=4, sw=25, n_hidden=128, n_layers=1, n_heads=4, slice_num=64, mlp_ratio=2, wd=0.0001, huber_delta=0.01
6. Use `--wandb_name alphonse/onecycle-100ep` and `--wandb_group lr-schedule`

## Baseline

Current best (`edward/lr6e3-sw25-100ep`):
- surf_p=33.55, surf_ux=0.4875, surf_uy=0.2704
- vol_p=63.80, best_epoch=97, val_loss=0.0190
- Config: lr=0.006, surf_weight=25, epochs=100, n_hidden=128, n_layers=1, n_heads=4, slice_num=64, mlp_ratio=2, batch_size=4, weight_decay=0.0001, huber_delta=0.01

---
## Results

**W&B run:** `3cbql15a` (alphonse/onecycle-100ep, group: lr-schedule)

**Training:** 40 epochs completed before 5-minute wall-clock timeout (hit at 5.1 min). Best at epoch 33.

**Peak memory:** 4.3 GB

| Metric | This run (OneCycleLR) | Baseline (CosineAnneal) | Delta |
|--------|----------------------|------------------------|-------|
| val/loss | 2.9403 | 0.0190 | ~150x worse (scale differs: MSE vs Huber) |
| surf_p | 123.4 | 33.55 | +268% worse |
| surf_Ux | 1.52 | 0.4875 | +212% worse |
| surf_Uy | 0.96 | 0.2704 | +255% worse |
| vol_p | 179.2 | 63.80 | +181% worse |
| best_epoch | 33 | 97 | — |

**What happened:**

OneCycleLR performed significantly worse than the baseline across all surface metrics. Two confounding factors complicate the comparison:

1. **Loss function mismatch:** The baseline uses Huber loss (delta=0.01), but this branch uses plain MSE. This explains the enormous val/loss scale difference (2.94 vs 0.019) and could also affect training stability at high LR.

2. **Timeout cuts off the schedule:** OneCycleLR is designed for full-budget training. The schedule peaks around epoch 5 (5% warmup) and then decays. At epoch 40 (the timeout boundary), we're still in the early decay phase — the LR was 0.0042 at the end. The deep final decay (to max_lr/25/1e4 ≈ 9.6e-8) that gives OneCycleLR its fine-convergence advantage never happens. The baseline (best_epoch=97) needed the full budget; OneCycleLR needs it even more.

3. **n_layers=1** means the model is very shallow — the baseline config matches this, so it's a wash, but it limits absolute performance ceiling.

In short: with a 5-minute ceiling (~40 epochs), OneCycleLR's warmup-then-decay schedule cannot demonstrate its promised advantage. The hypothesis may still be sound for uncapped training, but within the experimental budget it underperforms.

**Suggested follow-ups:**
- Add Huber loss (delta=0.01) to isolate scheduler effect from loss function effect — the current comparison isn't apples-to-apples
- Try `pct_start=0.02` or `pct_start=0.01` to shorten warmup and leave more decay budget within the timeout window
- Try a budget-aware scheduler: CosineAnnealing with T_max=40 (matching actual epoch count instead of 100) — prior failures may have used T_max=100 on a 50-epoch budget, which is a different mismatch

### CODE REVIEW

**Comment by tcapelle** [comment]
## Advisor Review — Closing

**Result:** surf_p=123.43 vs baseline 33.55. Severe regression.

**Analysis:** Multiple confounding factors: (1) MSE vs Huber mismatch — student used MSE while baseline uses Huber, (2) Only 33 epochs completed due to GPU contention, (3) OneCycleLR needs full budget to show its advantage. The student's analysis is thorough and correct.

**Takeaway:** OneCycleLR can't be tested fairly within 5-min + GPU contention constraints. The MSE/Huber mismatch is also a recurring problem — future experiments must use the exact same loss function as the baseline.
